### PR TITLE
Support for stmtListExpr in parser after major keywords

### PR DESCRIPTION
**Pull request by cooldome**
Support for stmtListExpr in parser after keywords. Makes parser and renderer closer to each other. Allows parsing code rendered after template expansion that produces nkStmtListExpr 

~~The change is not 100% backwards compatible, but is 99.9% compatible ;).~~
With added fix for #4035 managed to get full compatibility.

### CODE REVIEW

**Comment by cooldome** [comment]
No tests failed, I will continue with tests

**Comment by cooldome** [comment]
I have added a fix for high priority item into the mix: #4035. with this fix I managed to get full backwards compatibility. At least no failures known failures.
Changelog update.

Ready for review

**Comment by krux02** [comment]
Is this change necessary to fix #4035? Because I don't see how this change and issue #4035 is related.

I don't like this change, because I cannot predict it's consequences. I would limit parser changes to a bare minimum and here I don't see the need to do so. To get parser and renderer closer to each other, I would prefer if you would fix the renderer. For example render could render something like this:

```nim
proc c(x: int): int = 
  if (##[ "a comment, this block mimics the template expansion" ]##; let y = x+2; y>0):
    return y
```

Yes it looks ugly, but it is an already legal way to write a stmtListExpr after an ``if``.

**Comment by cooldome** [comment]
To be clear how  #4035 came in to the picture. I have found only one regression caused by my changes before fixing  #4035. The following no longer compiled:
```nim
return 
  a + 
  b
```
with my changes, it had to be rewritten as
```nim
return a + 
  b
```
or
```nim
return
  a + 
    b
```
with #4035 fixed there are no more known regressions.

**Comment by cooldome** [comment]
I am aware of `(##[ "a comment, this block mimics the template expansion" ]##; let y = x+2; y>0)` syntax.
I was looking for a way to make this example compilable:
```nim
  raise 
    var e: ref Exception
    new(e)
    e.msg = "My Exception msg"
    e
```
Which is very neat in my opinion. Once I did it, I have found that it is also easy to extend to `if, elif, case` 
 conditions despite of semicolon expectation at the end of expression.

In case anyone will find any relevant problems with this PR. I can scale down this change to `raise, return, discard, yield` where it makes sense 100%.

**Comment by Araq** [comment]
I think it's a fine patch but new features should be put under an experimental switch, ``experimental: exprStmtListExtensions``

File: tests/parser/tstmtlist_expr.nim [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+proc b(x: int): int = 
+  raise 
+    var e: ref Exception
+    new(e)
+    e.msg = "My Exception msg"
+    e
+
+proc c(x: int): int = 
+  if
```
Comment (zah) [inline tests/parser/tstmtlist_expr.nim]: There is no if condition here.

Comment (zah) [reply tests/parser/tstmtlist_expr.nim]: Ah, sorry, I understand the example now.

**Comment by cooldome** [comment]
I have added a experimental switch `exprStmtListExtensions`. Unfortunately, the following doesn't work
```nim
{.experimental: "exprStmtListExtensions"}.
```
As parser advances faster than sempass process the pragma block. Feature needs to be enabled in command line, config or nimscript.

**Comment by Araq** [comment]
Ok, rejecting this for now. renderer.nim should use ``(;)`` in if/case/etc conditions that produce ``nkStmtListExpr``.

**Pull request closed without merging**